text/0000-build-std.md

```diff
@@ -0,0 +1,3601 @@
+- [*Why not produce a `dylib` for the standard library?*][why-not-produce-a-dylib-for-the-standard-library]
+- [*Why also build `proc_macro`, `test` and `profiler_builtins`?*][why-also-build-proc_macro-test-and-profiler_builtins]
+- [*Why rebuild the standard library automatically?*][why-rebuild-the-standard-library-automatically]
+- [*Why use the lockfile of the `rust-src` component?*][why-use-the-lockfile-of-the-rust-src-component]
+
+### Profiles
+[profiles]: #profiles
+
+Cargo will assume that the pre-built standard library matches the default
+release profile. If the user changes the default release profile or builds with
+a different profile then depending on the value of `build-std` in the
+`.cargo/config`, this could trigger a rebuild. rustc will add a
+`--print target-modifiers` flag which will print all of the flags treated as
+target modifiers (e.g. one flag per line, like `-Zretpoline`). If changing a
+profile setting would result in one of these flags being emitted by Cargo, then
+a target modifier has changed and would no longer match the pre-built standard
+library.
+
+Standard library crates will be built using the configuration of the current
+profile defined in the standard library's workspace. For example, if building in
+the release profile, the release profile of the standard library workspace will
+be used. If the profile of the user's crate sets a configuration option, that
+will be merged with the standard library's profile (e.g. if the user sets
+`profile.release opt-level` that will override the standard library's release
+`opt-level`, or if the user sets `profile.release.rustflags` that will be
+appended to the standard library's release `rustflags`).
+
+Profile overrides in the standard library's workspace continue to apply to its
+dependencies. User profile overrides can only apply to the `std`, `alloc` and
+`core` crates. 
+
+**TODO:** discuss moving config into profile from bootstrap
+
+*See the following sections for rationale/alternatives:*
+- [*Why not always rebuild when the profile changes?*][why-not-always-rebuild-when-the-profile-changes]
+- [*Why not ship a debug profile `rust-std`?*][why-not-ship-a-debug-profile-rust-std]
+- [*Why respect profile overrides of the standard library's workspace?*][why-respect-profile-overrides-of-the-standard-librarys-workspace]
+
+### Preventing implicit sysroot dependencies
+[preventing-implicit-sysroot-dependencies]: #preventing-implicit-sysroot-dependencies
+
+Cargo will pass a new `--no-implicit-sysroot-deps` flag to rustc which will
+prevent rustc from loading top-level dependencies from the sysroot. For example,
+writing `extern crate foo` in a crate will not load `foo.rlib` from the sysroot
+if it is present, but if an `--extern noprelude:bar.rlib` is provided which
+depends on a crate `foo`, rustc will look in `-L` paths and the sysroot for it.
```

Comment:
Why is it necessary for rust to look in the sysroot for deps? For "Special object files", can't the linker bring these in?

Reply:
The idea behind this part was that when build-std does a rebuild, `std` is provided via `--extern` and the standard library's deps are in the normal `target/deps` directory and found via `-L`, but when using `--extern` with the sysroot `std`, then the deps need to be in the search paths somewhere, so the sysroot still needs to be in the search path. I don't think this is related to the special object files.

Reply:
I was worried about race conditions here a la https://github.com/rust-lang/wg-cargo-std-aware/issues/49, but after thinking about it for a bit I can't think of any scenario that I think would break since your direct dependencies are always explicit